docs/design/tools/illink/feature-switch-attributes.md

```diff
@@ -313,39 +311,115 @@ However, any attribute-based model that we pick to represent features should be
 We would like for the attribute model to have a consistent way to refer to a feature in `FeatureGuardAttribute` or `FeatureSwitchAttribute`. The proposal is to support an attribute model for features that have an associated `Requires` attribute, and use that attribute type uniformly to refer to the feature from `FeatureGuardAttribute` and `FeatureSwitchAttribute`.
 
 For example, the feature switch for "dynamic code support" might look like this:
+
 ```csharp
 public class RuntimeFeature
 {
-    [FeatureSwitch<RequiresDynamicCode>]
+    [FeatureSwitch(typeof(RequiresDynamicCodeAttribute))]
     public static bool IsDynamicCodeSupported => // ...
 }
 
-class RequiresDynamicCodeAttribute : Attribute, IFeatureAttribute<RequiresDynamicCodeAttribute> {
-    public static string FeatureName => "System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported";
-}
+
+```csharp
+[FeatureName("System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported")]
+class RequiresDynamicCodeAttribute : FeatureAttribute { }
 ```
+
 and a feature guard for "dynamic code compilation" might look like this:
 ```csharp
 public class RuntimeFeature
 {
-    [FeatureGuard<RequiresDynamicCode>]
-    public static bool IsDynamicCodeSupported => // ...
+    [FeatureGuard(typeof(RequiresDynamicCodeAttribute))]
+    public static bool IsDynamicCodeCompiled => // ...
 }
 ```
 
 The attribute definitions to support this might look like this:
 ```csharp
-class FeatureGuardAttribute<T> : Attribute
-  where T : Attribute, IFeatureAttribute<T> {}
+[AttributeUsage(AttributeTargets.Class, Inherited = false)]
+abstract class FeatureAttribute : Attribute { }
```

Comment:
Worth mentioning - this would effectively provide a way to define custom `Requires...` attributes. Something we've discussed several times in the past. We should note that it's not part of this proposal to actually make that work. This will lead to somewhat weird place where if I define my own feature using these attributes, I would have the guard and everything work correctly, but I still would not get any warnings calling the `Requires..` annotated methods outside of the checked scope.